src/types/Slot0.sol

```diff
@@ -0,0 +1,98 @@
+pragma solidity ^0.8.20;
+
+/** @dev Slot0 is a packed version of solidity structure. 
+ * Using the packaged version saves gas by not storing the structure fields in memory slots.
+ * 
+ * Fields:
+ * 
+ * The current price
+ * uint160 sqrtPriceX96;
+ * 
+ * The current tick
+ * int24 tick;
+ * 
+ * Protocol fee, expressed in hundredths of a bip, upper 12 bits are for 1->0, and the lower 12 are for 0->1
+ * the maximum is 1000 - meaning the maximum protocol fee is 0.1%
+ * the protocolFee is taken from the input first, then the lpFee is taken from the remaining input
+ * uint24 protocolFee;
+ * 
+ * Used for the lp fee, either static at initialize or dynamic via hook
+ * uint24 lpFee;
+ */
+type Slot0 is bytes32;
+
+using Slot0Library for Slot0 global;
+
+library Slot0Library {
+    uint160 private constant UINT160_MASK = 0x00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
+    uint24 private constant INT24_MASK = 0xFFFFFF;
+    uint24 private constant UINT24_MASK = 0xFFFFFF;
+
+    uint8 private constant TICK_OFFSET = 160;
+    uint8 private constant PROTOCOL_FEE_OFFSET = 184;
+    uint8 private constant LP_FEE_OFFSET = 208;
+
+    // #### GETTERS ####
+    function sqrtPriceX96(Slot0 _packed) internal pure returns (uint160 _sqrtPriceX96) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _sqrtPriceX96 := and(UINT160_MASK, _packed)
+        }
+    }
+
+    function tick(Slot0 _packed) internal pure returns (int24 _tick) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _tick := signextend(2, shr(TICK_OFFSET, _packed))
+        }
+    }
+
+    function protocolFee(Slot0 _packed) internal pure returns (uint24 _protocolFee) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _protocolFee := and(UINT24_MASK, shr(PROTOCOL_FEE_OFFSET, _packed))
+        }
+    }
+
+    function lpFee(Slot0 _packed) internal pure returns (uint24 _lpFee) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _lpFee := and(UINT24_MASK, shr(LP_FEE_OFFSET, _packed))
+        }
+    }
+
+    // #### SETTERS ####
+    function setSqrtPriceX96(Slot0 _packed, uint160 _sqrtPriceX96) internal pure returns (Slot0 _result) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _result := or(and(not(UINT160_MASK), _packed), and(UINT160_MASK, _sqrtPriceX96))
+        }
+    }
+
+    function setTick(Slot0 _packed, int24 _tick) internal pure returns (Slot0 _result) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _result := or(and(not(shl(TICK_OFFSET, INT24_MASK)), _packed), shl(TICK_OFFSET, and(INT24_MASK, _tick)))
+        }
+    }
+
+    function setProtocolFee(Slot0 _packed, uint24 _protocolFee) internal pure returns (Slot0 _result) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            _result :=
+                or(
+                    and(not(shl(PROTOCOL_FEE_OFFSET, UINT24_MASK)), _packed),
```

Comment:
presumably optimiser is catching that `not(shl(PROTOCOL_FEE_OFFSET, UINT24_MASK))`  is always the same?

Reply:
Will check it

Reply:
Indeed, attempts to use constants instead of `not(shl(...))` do not change anything or even make gas worse